aten/src/THC/THCTensorIndex.cu

```diff
@@ -66,21 +66,30 @@ __global__ void indexCopySmallIndex(TensorInfo<T, IndexType> dst,
 // the number of indices chosen is small, then the
 // indexCopySmallIndex kernel is a better choice to reduce memory
 // accesses.
-template <typename T, typename IndexType, int DstDim, int SrcDim, int IdxDim>
+template <typename T, typename IndexType, int DstDim, int SrcDim, int IdxDim,
+          bool IndexIsMajor>
 __global__ void indexCopyLargeIndex(TensorInfo<T, IndexType> dst,
                                     TensorInfo<T, IndexType> src,
                                     TensorInfo<int64_t, IndexType> indices,
                                     int dstCopyDim,
                                     int srcCopyDim,
+                                    IndexType totalSize,
                                     IndexType innerSize,
                                     int64_t dstCopyDimSize) {
   // We stride over the output including the indexed dimension
   // (totalSize), and calculate the destination index point based on that
   for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
-       linearIndex < innerSize * indices.sizes[0];
+       linearIndex < totalSize;
        linearIndex += gridDim.x * blockDim.x) {
-    IndexType srcIndex = linearIndex / innerSize;
-    IndexType elementInSlice = linearIndex % innerSize;
+    IndexType srcIndex, elementInSlice;
+    if (IndexIsMajor) {
+      srcIndex = linearIndex / innerSize;
+      elementInSlice = linearIndex % innerSize;
+    }
+    else {
+      elementInSlice = linearIndex / innerSize;
+      srcIndex = linearIndex % innerSize;
```

Comment:
If you're looking for more speedups in these kernels then removing this division and modulo will go a long way. I've seen that have a huge impact on (Log)Softmax kernels I wrote. A 2D grid with bounds check ended up being much faster.

Reply:
Hmm... that's interesting, could you point me to your change and tell me what hardware you used?

I recently toyed with reducing the number of the integer divisions in IndexToOffset for pointwise operations, but could barely see any speedup.  It looked like pointwise operations are dominated by memory latency, at least for my GTX 1080.

Reply:
Integer multiplication and addition is fine - division and modulo is on the order of 10-100x slower IIRC. This is a [function that computes a 2d block size](https://github.com/pytorch/pytorch/pull/3245/files#diff-17f33cbf97c7a067866a1408f2c2926eR32), and one that computes the grid size is below. You can see how this is later [used in the kernel itself](https://github.com/pytorch/pytorch/pull/3245/files#diff-17f33cbf97c7a067866a1408f2c2926eR109).

Reply:
BTW, this idea is very interesting, but it's an independent change so probably better handled in a separate PR.  Can't make promises, but I might visit it if I have more spare time.  :)